util/patch/patch.go

```diff
@@ -18,117 +18,292 @@ package patch
 
 import (
 	"context"
+	"encoding/json"
 	"reflect"
+	"time"
 
 	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
 	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/cluster-api/util/conditions"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// Helper is a utility for ensuring the proper Patching of resources
-// and their status
+// Helper is a utility for ensuring the proper patching of objects.
 type Helper struct {
-	client        client.Client
-	before        map[string]interface{}
-	hasStatus     bool
-	beforeStatus  interface{}
-	resourcePatch client.Patch
-	statusPatch   client.Patch
+	opts *HelperOptions
+
+	client       client.Client
+	beforeObject runtime.Object
+	before       *unstructured.Unstructured
+	after        *unstructured.Unstructured
+	changes      map[string]bool
+
+	canInterfaceConditions bool
```

Comment:
as discussed earlier, consider renaming `canInterfaceConditions` to something more clear. Maybe `conditionsEnabled` or `hasConditions`?

Reply:
handleConditionsConcurrency?

Reply:
Renamed this to `isConditionsSetter`